rust/main/lander/src/tests/ethereum/inclusion_stage.rs

```diff
@@ -124,32 +97,109 @@ async fn test_inclusion_gas_spike() {
+            elapsed,
+            base_processing_delay,
+            inclusion_stage_processing_delay,
+            block_time_clone,
+            &tx,
+            // First submission, price is 200000 - the default fee used by the ethers estimation logic
+            // Second submission, price is 10% higher, even though the spike was smaller
+            // Third submission, price is 10% higher again, even though the spike was smaller
+            // Fourth submission, price matches the spike, because it was greater than 10%
+            vec![200000, 220000, 242000, 300000],
+        );
         Ok(H256::random()) // Mocked transaction hash
     });
 
     mock_evm_provider
         .expect_get_block()
         .returning(|_| Ok(Some(mock_block(42, 100)))); // Mocked block retrieval
 
+    run_and_expect_successful_inclusion(mock_evm_provider, block_time).await;
+}
+
+#[tokio::test]
+#[traced_test]
+async fn test_inclusion_gas_underpriced() {
+    let block_time = Duration::from_millis(20);
+    let mut mock_evm_provider = MockEvmProvider::new();
+    mock_evm_provider
+        .expect_get_finalized_block_number()
+        .returning(|_reorg_period| {
+            Ok(43) // Mocked block number
+        });
+    mock_evm_provider
+        .expect_estimate_gas_limit()
+        .returning(|_, _| {
+            Ok(21000.into()) // Mocked gas limit
+        });
+
+    mock_evm_provider
+        .expect_fee_history()
+        .returning(move |_, _, _| Ok(mock_fee_history(200000, 10)));
+
+    // after the tx is sent and gets a tx hash, immediately report it as included
+    mock_evm_provider
+        .expect_get_transaction_receipt()
+        .returning(move |_| Ok(Some(mock_tx_receipt(Some(42)))));
+
+    // assert each expected price by mocking the `send` method of the provider
+    let mut send_call_counter = 0;
+    let elapsed = Instant::now();
+    let base_processing_delay = Duration::from_millis(15);
+    // assume 1 second more than usual because that's the retry delay when an error occurs
+    let inclusion_stage_processing_delay = Duration::from_millis(1010);
+    let block_time_clone = block_time.clone();
+    mock_evm_provider.expect_send().returning(move |tx, _| {
+        send_call_counter += 1;
+        // assert the timing of resubmissions to make sure they don't happen more than once per block
+        assert_gas_prices_and_timings(
+            send_call_counter,
+            elapsed,
+            base_processing_delay,
+            inclusion_stage_processing_delay,
+            block_time_clone,
+            &tx,
+            // First submission, price is 200000 - the default fee used by the ethers estimation logic
+            // Second submission, price is 10% higher, even though the spike was smaller
+            // Third submission, price is 10% higher again, even though the spike was smaller
+            // Fourth submission, price matches the spike, because it was greater than 10%
+            vec![200000, 220000],
```

Comment:
I suppose the comment is incorrect since we expect only 2 submission attempts here.

Reply:
ah good catch, for this test we only need a single escalation. This is outdated copy pasta